relay-server/src/services/spooler/envelope_stack/sqlite.rs

```diff
@@ -0,0 +1,684 @@
+            // in a critical state, so we return an error.
+            if let Some(db_error) = db_error {
+                return Err(SQLiteEnvelopeStackError::DatabaseError(db_error));
+            }
+
+            // In case no envelopes were unspool, we will mark the disk as empty until another round
+            // of spooling takes place.
+            self.check_disk = false;
+
+            return Ok(());
+        }
+
+        // We sort envelopes by `received_at`.
+        extracted_envelopes.sort_by_key(|a| received_at(a));
+
+        // We push in the back of the buffer, since we still want to give priority to
+        // incoming envelopes that have a more recent timestamp.
+        self.batches_buffer_size += extracted_envelopes.len();
+        self.batches_buffer.push_back(extracted_envelopes);
+
+        Ok(())
+    }
+
+    /// Deserializes an [`Envelope`] from a database row.
+    fn extract_envelope(&self, row: SqliteRow) -> Result<Box<Envelope>, SQLiteEnvelopeStackError> {
+        let envelope_row: Vec<u8> = row
+            .try_get("envelope")
+            .map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+        let envelope_bytes = bytes::Bytes::from(envelope_row);
+        let mut envelope =
+            Envelope::parse_bytes(envelope_bytes).map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+
+        let received_at: i64 = row
+            .try_get("received_at")
+            .map_err(|_| SQLiteEnvelopeStackError::Empty)?;
+        let start_time = StartTime::from_timestamp_millis(received_at as u64);
+
+        envelope.set_start_time(start_time.into_inner());
+
+        Ok(envelope)
+    }
+
+    /// Validates that the incoming [`Envelope`] has the same project keys at the
+    /// [`SQLiteEnvelopeStack`].
+    fn validate_envelope(&self, envelope: &Envelope) -> bool {
+        let own_key = envelope.meta().public_key();
+        let sampling_key = get_sampling_key(envelope).unwrap_or(own_key);
+
+        self.own_key == own_key && self.sampling_key == sampling_key
+    }
+}
+
+impl EnvelopeStack for SQLiteEnvelopeStack {
+    type Error = SQLiteEnvelopeStackError;
+
+    async fn push(&mut self, envelope: Box<Envelope>) -> Result<(), Self::Error> {
+        if !self.validate_envelope(&envelope) {
+            return Err(Self::Error::MismatchedEnvelope);
+        }
+
+        if self.above_spool_threshold() {
+            self.spool_to_disk().await?;
+        }
+
+        // We need to check if the topmost batch has space, if not we have to create a new batch and
+        // push it in front.
+        if self
+            .batches_buffer
+            .front()
+            .map_or(true, |last_batch| last_batch.len() >= self.batch_size.get())
+        {
+            self.batches_buffer
+                .push_front(Vec::with_capacity(self.batch_size.get()));
+        }
+
+        if let Some(batch) = self.batches_buffer.front_mut() {
+            batch.push(envelope);
+            self.batches_buffer_size += 1;
+        }
```

Comment:
nit: Instead of checking `front_mut` again here, I think we can either `.expect()` it, or 

```rust
if let Some(front) = self.batches_buffer.front().filter(|b| b.len() < batch_size) {
    front.push(envelope);
} else {
    self.batches_buffer.push_front(vec![envelope]); // omitted capacity here out of laziness
}
```

Reply:
Yeah definitely better!